x-element.js

```diff
@@ -1012,25 +1012,70 @@ export default class XElement extends HTMLElement {
 
 /** Internal implementation details for template engine. */
 class TemplateEngine {
-  static #UNSET = Symbol(); // Ensures a unique, initial comparison.
-  static #ATTRIBUTE_PREFIXES = {
-    attribute: 'x-element-attribute-',
-    boolean: 'x-element-boolean-',
-    property: 'x-element-property-',
-  };
-  static #CONTENT_PREFIX = 'x-element-content-';
-  static #CONTENT_REGEX = new RegExp(`${TemplateEngine.#CONTENT_PREFIX}(\\d+)`);
-  static #OPEN = /<[a-z][a-z0-9-]*(?=\s)/g;
-  static #STEP = /\s+[a-z][a-z0-9-]*(?=[\s>])|\s+[a-z][a-zA-Z0-9-]*="[^"]*"/y;
-  static #ATTRIBUTE = /\s+(\??([a-z][a-zA-Z0-9-]*))="$/y;
-  static #PROPERTY = /\s+\.([a-z][a-zA-Z0-9_]*)="$/y;
-  static #CLOSE = />/g;
-
-  static #interface = null;
-  static #stateMap = new WeakMap(); // Maps nodes to internal state.
-  static #analysisMap = new WeakMap(); // Maps strings to cached computations.
-  static #resultMap = new WeakMap(); // Maps symbols to results.
-  static #updaterMap = new WeakMap(); // Maps symbols to updaters.
+  // Special markers added to markup enabling discovery post-instantiation.
+  static #ATTRIBUTE_PREFIX = 'x-element-attribute';
+  static #BOOLEAN_PREFIX = 'x-element-boolean';
+  static #DEFINED_PREFIX = 'x-element-defined';
+  static #PROPERTY_PREFIX = 'x-element-property';
+  static #CONTENT_PREFIX = 'x-element-content';
+
+  // Patterns to find special edges in original html strings.
+  static #OPEN_REGEX = /<[a-z][a-z0-9-]*(?=\s)/g;
+  static #STEP_REGEX = /(?:\s+[a-z][a-z0-9-]*(?=[\s>])|\s+[a-z][a-zA-Z0-9-]*="[^"]*")+/y;
+  static #ATTRIBUTE_OR_PROPERTY_REGEX = /\s+(?:(?<questions>\?{0,2})?(?<attribute>([a-z][a-zA-Z0-9-]*))|\.(?<property>[a-z][a-zA-Z0-9_]*))="$/y;
+  static #CLOSE_REGEX = />/g;
+
+  // Sentinel to initialize the “last values” array.
+  static #UNSET = Symbol();
+
+  // Mapping of container nodes to internal state.
+  static #nodeToState = new WeakMap();
+
+  // Mapping of starting comment cursors to internal array state.
+  static #nodeToArrayState = new WeakMap();
+
+  // Mapping of tagged template function “strings” to caches computations.
+  static #stringsToAnalysis = new WeakMap();
+
+  // Mapping of opaque references to internal result objects.
+  static #symbolToResult = new WeakMap();
+
+  // Mapping of opaque references to internal update objects.
+  static #symbolToUpdate = new WeakMap(); 
```

Comment:
I improved the naming of all the maps in this file. They’re just more explicit now.